Support switching RemoteModule between train/eval

lishanYang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59026 Support switching RemoteModule between train/eval**

#Closes: https://github.com/pytorch/pytorch/issues/51480

Enabled methods train and eval in RemoteModule to call the underlying train/eval methods on the actual
 nn.Module

Differential Revision: [D28721078](https://our.internmc.facebook.com/intern/diff/D28721078/)

Commit:
Support switching RemoteModule between train/eval

Enabled methods train and eval in RemoteModule to call the underlying train/eval methods on the actual
 nn.Module

Differential Revision: [D28721078](https://our.internmc.facebook.com/intern/diff/D28721078/)

[ghstack-poisoned]

SciPioneer commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -505,6 +504,16 @@ def test_send_remote_module_over_the_wire(self):
                 dst_worker2_name, remote_forward_async, (remote_module, args)
             )
             self.assertEqual(ret2, tuple(reversed(args)))
+            
```
Can you move these to 2 new test cases above `test_unsupported_methods`? It's better to have different test cases specific to different functionality. You can see that `get_module_rref` and `remote_parameters()` methods are tested separately.

SciPioneer commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -422,10 +422,10 @@ def named_modules(
 
     def train(self: T, mode: bool = True) -> T:  # type: ignore[return]
-        _raise_not_supported(self.train.__name__)
+        return self.module_rref.rpc_sync().train()
```
This will make `train()` a blocking op. Not sure if we need to have a separate async version.

@pbelevich Do you think we need an async version? I am not sure if a local host always wants to wait for the training on a remote worker.

SciPioneer commented:
Overall looks good! Please address the comment on the unit tests.

Please also check with other people if there is a need to support async train/eval. Thanks!

pritamdamania87 commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -505,6 +504,16 @@ def test_send_remote_module_over_the_wire(self):
             )
             self.assertEqual(ret2, tuple(reversed(args)))
+            
+            args = True
+            ret3 = rpc.rpc_sync(dst_worker2_name, remote_train, (remote_module, args))
+            self.assertEqual(ret3.training, True)
+            
+            ret4 = rpc.rpc_sync(dst_worker2_name, remote_eval, (remote_module,))
+            self.assertEqual(ret4.training, False)
```
Probably add a separate test method for this instead of using this. Also, what you probably want to test is something like this:

1. Run `remote_module.train()` and `remote_module.eval()` locally
2. Use RPC to retrieve the training parameter remotely, you can define a function like this and call it over RPC:

```
def get_training_arg(module_rref):
  return module_rref.local_value().training

# Call it
rpc.rpc_sync(dst, get_training_arg, args=(remote_module.get_module_rref()))

Commit:
Update on "Support switching RemoteModule between train/eval"


#Closes: https://github.com/pytorch/pytorch/issues/51480

Enabled methods train and eval in RemoteModule to call the underlying train/eval methods on the actual
 nn.Module

Differential Revision: [D28721078](https://our.internmc.facebook.com/intern/diff/D28721078/)

[ghstack-poisoned]

pritamdamania87 commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -257,6 +260,32 @@ def test_get_module_rref(self):
             for param in rref.to_here().parameters():
                 self.assertTrue(torch.equal(param, _PARAM_VAL))
 
+    @dist_utils.dist_init
+    def test_train(self):
+        if self.rank != 0:
+            return
+        dst_worker_name = dist_utils.worker_name((self.rank + 1) % self.world_size)
+
+        for remote_module in self._create_remote_module_iter(
+            dst_worker_name, modes=[ModuleCreationMode.MODULE_CTOR]
+        ):
+            remote_module.train()
+            ret = rpc.rpc_sync(dst_worker_name, get_remote_training_arg, args=(remote_module.get_module_rref(),))
+            self.assertEqual(ret, True)
+
+    @dist_utils.dist_init
+    def test_eval(self):
+        if self.rank != 0:
+            return
+        dst_worker_name = dist_utils.worker_name((self.rank + 1) % self.world_size)
+
+        for remote_module in self._create_remote_module_iter(
+            dst_worker_name, modes=[ModuleCreationMode.MODULE_CTOR]
+        ):
+            remote_module.eval()
+            ret = rpc.rpc_sync(dst_worker_name, get_remote_training_arg, args=(remote_module.get_module_rref(),))
+            self.assertEqual(ret, False)
```
nit: Can combine these both into a single test method `test_train_eval` since there is quite a bit of duplicated code.

SciPioneer replied:
+1

Commit:
Update on "Support switching RemoteModule between train/eval"


#Closes: https://github.com/pytorch/pytorch/issues/51480

Enabled methods train and eval in RemoteModule to call the underlying train/eval methods on the actual
 nn.Module

Differential Revision: [D28721078](https://our.internmc.facebook.com/intern/diff/D28721078/)

[ghstack-poisoned]

Commit:
Update on "Support switching RemoteModule between train/eval"


#Closes: https://github.com/pytorch/pytorch/issues/51480

Enabled methods train and eval in RemoteModule to call the underlying train/eval methods on the actual
 nn.Module

Differential Revision: [D28721078](https://our.internmc.facebook.com/intern/diff/D28721078/)

[ghstack-poisoned]

Commit:
Update on "Support switching RemoteModule between train/eval"


#Closes: https://github.com/pytorch/pytorch/issues/51480

Enabled methods train and eval in RemoteModule to call the underlying train/eval methods on the actual
 nn.Module

Differential Revision: [D28721078](https://our.internmc.facebook.com/intern/diff/D28721078/)

[ghstack-poisoned]

Commit:
Update on "Support switching RemoteModule between train/eval"


#Closes: https://github.com/pytorch/pytorch/issues/51480

Enabled methods train and eval in RemoteModule to call the underlying train/eval methods on the actual
 nn.Module

Differential Revision: [D28721078](https://our.internmc.facebook.com/intern/diff/D28721078/)

[ghstack-poisoned]
